Fixes bug 1248928 - Experimental new home page based on the public API.

adngdb opened the pull request:
cc @lonnen 

Let's discuss whether this is a good idea or not. As far as I can tell it has the same features as the current home page, minus a few bugs. It makes 3 HTTP requests to get the data (and 2 more if you change the duration since we only call the ``ProductBuildTypes`` service once) instead of 1 for the current home. The logic is in a single JS file of less than 400 lines with comments (the logic of the current home page is, well, quite complex, with lots of function calls, passing data around... ).

I have not built this with the crashes per day page in mind, so if we want to share some code between those two pages we'll have to do some refactoring.

peterbe commented:
Remember; we need to solve the case where a product has **no ADI** for any of its versions.

peterbe commented on webapp-django/crashstats/crashstats/templatetags/jinja_helpers.py:
```diff
@@ -92,6 +92,11 @@ def json_dumps(data):
 
+@library.filter
+def to_json(data):
+    return json.dumps(data)
```
I don't like this. It's dangerous. If you do something like this in a jinja template:
```
<script>var DATA = {{ some_view_python_data | to_json }};</script>
```
**and** somewhere inside that `some_view_python_data` there's a `</script>` it'll actually render. E.g. Suppose that you change your name from "Adrian" to `</script><script>alert('xss')</script>`, then you'd get this:
```html
<script>var DATA = {"name": "</script><script>alert('xss')</script>"};</script>
```
and you have an XSS hole.

peterbe replied:
I know we're not using this `to_json` to in a inline script block. We're using inside a data tag. However, there might be other vulnerabilities to do with data tags. Or perhaps we forget and use it in some dangerous way in some other jinja template.

peterbe replied:
The solution is to do this:
```python
def to_json(data):
    return json.dumps(data).replace('</', '<\\/')
```

peterbe commented:
Are you aware how different they look?
<img width="1816" alt="screen shot 2016-02-22 at 3 48 49 pm" src="https://cloud.githubusercontent.com/assets/26739/13232136/d586c768-d97b-11e5-897b-07d9a7da09f3.png">
<img width="1816" alt="screen shot 2016-02-22 at 3 49 26 pm" src="https://cloud.githubusercontent.com/assets/26739/13232142/df69cfb4-d97b-11e5-98b9-ed381dc3a6d6.png">

peterbe commented:
By the way, looking at the new graph, it appears like there was no crashes on Feb 21 (which was "yesterday" at the time of writing). But there were.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+    // Versions ending in -b are "magic" versions that will be split into all
+    // of the beta versions starting with that same version number.
+    var betaVersions = [];
+    for (var i = versions.length - 1; i >= 0; i--) {
```
Why not `versions.forEach`?

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+    endDate.setDate(endDate.getDate() - 1);  // We start with yesterday.
+    var startDate = new Date();
+    startDate.setDate(endDate.getDate() - duration);
+
```
I was allowed mess with the duration to some number that is NOT one of the duration options
<img width="1816" alt="screen shot 2016-02-22 at 4 20 18 pm" src="https://cloud.githubusercontent.com/assets/26739/13233140/49a7ec4a-d980-11e5-9960-e18999fabfa1.png">
Should we not guard against that?

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+        var days = e.state.duration;
+
+        $('#duration .selected').removeClass('selected');
+        $('#duration .days-' + days).addClass('selected');
```
Here we should perhaps do something like:
```javascript
if (!$('#duration .days-' + days).length) {
    alert("Invalid duration choice");
}
```

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+     * Click event handler for the duration menu. Changes the history and
+     * reloads the graph's data.
+     */
+    $('#duration a').click(function (e) {
```
Change this to
```javascript
$('#duration').on('click', 'a', function(e) {
```
That's more performant. It's better to have 1 event handler listening and within it, delegate based on what was clicked.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+            url: url,
+            success: onSuperSearchData,
+            error: onQueryError,
+        });
```
Cool!

I'm almost wondering if we should hardcode the value for `superSearchApiUrl`. Since it's a fully exposed thing. A URL we're committed to not changing.

peterbe replied:
By the way, it has little functional difference in this context but I suggest you use promises instead. 
Unlike basic callbacks they can be chained. So if you change this to:
```
return $.ajax({url: url})
.done(onSuperSearchData)
.fail(onQueryError);
```
Careful if you rely on `this` on this functions. Then you might need to bind the original `this`.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+    /**
+     * Get count of average daily installs per day and per version.
+     */
+    function queryAdi() {
```
"adi" is not a word. It's an abbreviation. So, shouldn't this be `queryADI`?

adngdb replied:
I've read different styles about this. Some say that it should be ``UUID``, others ``Uuid``. I'm not sold on anything, so if you prefer (and already use) the full uppercase for abbreviations, I don't mind switching to that.

peterbe replied:
It's personal preference I guess. Personally I think I prefer `queryURL` or `productADI`.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+        // Make a simple structure to formalize the data, and make it easy to
+        // handle beta versions.
+        var dataStruct = {};
+        for (i = 0, ln = versions.length; i < ln; i++) {
```
Again, why not `versions.forEach`?

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+
+        var dates = data.facets.histogram_date;
+        for (i = 0, ln = dates.length; i < ln; i++) {
+            var date = dates[i];
```
ditto

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+
+                dataStruct[baseVersion][date.term.substring(0, 10)] += versionsCounts[j].count;
+            }
+        }
```
I read it but my head's spinning. What does this whole thing do? I know it's about filling up `dataStruct` but some human-readable summary would be nice.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
This needs to be entirely replaced with promises and deferreds. jQuery's built-in is quite nice. 
You're currently relying on `dataStore.superSearch` to exist. Which is kinda weird. Unlike Python, an array is always true. Even if there's nothing in there. Same for a dictionary. So if `dataStore.superSearch` is `[]` or `{}` this'll be true. 

I was going to say something stupid like "Check the documentation on $.when". But to back up my feedback, I put together a little test to make sure I understand this myself. 

http://codepen.io/peterbe/pen/wMLLwe?editors=0010

That way you get a promise call back once all the **post-processing** has been done **after** the AJAX has finished, for each AJAX query.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
Why not just `if (!dataStore.adi[day] || !dataStore.adi[day])`?

adngdb commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,386 @@
+    'use strict';
+
+    // A polyfill for `endsWith`
+    if (!String.prototype.endsWith) {
```
@peterbe What do you think about polyfills in Socorro? I guess you don't mind them since you didn't comment here. But should we maybe centralize them? I could see us having a file called ``polyfills.js`` that we would include on every page, and which would contain all of our polyfills. That way it's easy to maintain by regularly verifying if something can be removed.

peterbe replied:
Let's avoid globals as much as we possibly can. If a particular page needs a particular polyfill, we put in in that app's static dir. If we find we need them repeatedly, we move it to the `base`'s static dir. 

Global solutions are very hard to undo as the project grows. Exceptions might be jquery or d3 .

adngdb commented:
@peterbe This now works for products without ADIs. I have added validation for the duration. I have refactored the JS to use promises. And I have fixed the bug about the missing data for the last day of the covered time period.

adngdb commented on webapp-django/crashstats/settings/bundles.py:
```diff
@@ -228,8 +228,6 @@
             'crashstats/js/socorro/nav.js',
             'crashstats/js/socorro/analytics.js',
-            'browserid/api.js',
-            'browserid/browserid.js',
```
Haha, I can't believe we didn't catch this... :) (I was working on this in a train with no internet connection. )

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,462 @@
+
+        var url = adiApiUrl + '?' + $.param(params, true);
+        $.ajax({url: url})
+        .then(onADIData)
```
uh?! I didn't even know you can do that. I thought `$.ajax` returns a promise that only has `done` and `fail`.

peterbe commented on webapp-django/crashstats/home/static/home/js/home.js:
```diff
@@ -0,0 +1,462 @@
+
+        startDate = new Date();
+        startDate.setDate(endDate.getDate() - duration);
+    }
```
This is not working
http://codepen.io/peterbe/pen/EKaLWJ?editors=1010

For the record, if you look at this a day later and the logic breaks then. The startDate becomes March 22nd, when it's supposed to become FEB 22nd.

peterbe replied:
I suspect it's a leapyear bug https://twitter.com/peterbe/status/704641995147444224

peterbe commented:
r+ **but** I'd like to view it one last time once you've fixed the date arithmetic bug.

adngdb commented:
@peterbe forgot to mention it but the date problem has been addressed.

peterbe commented:
Still r+ :)

We'll launch it gently and test it on stage and prod with @m8ttyB and @KaiRo-at

peterbe commented:
FYI, @m8ttyB and @KaiRo-at
When Adrian lands this it's going to be https://crash-stats.allizom.org/new_home/products/Firefox  (just `home` replaced by `new_home`)
